### PR TITLE
Fixes #RHIROS-1269 - feature flags for backend

### PR DESCRIPTION
**Pull request by upadhyeammit**
## PR Title :boom:

Please title this PR with a summary of the change, along with the JIRA card number.

Suggested formats: 

1. Fixes/Refs #RHIROS-XXX - Title
2. RHIROS-XXX Title 

Feel free to remove this section from PR description once done.

## Why do we need this change? :thought_balloon:

Please include the __context of this change__ here.

## Documentation update? :memo:

- [ ] Yes
- [ ] No

## Security Checklist :lock:

Upon raising this PR please go through [RedHatInsights/secure-coding-checklist](https://github.com/RedHatInsights/secure-coding-checklist)

## :guardsman: Checklist :dart:

- [ ] Bugfix
- [ ] New Feature
- [ ] Refactor
- [ ] Unittests Added
- [ ] DRY code
- [ ] Dependency Added
- [ ] DB Migration Added

## Additional :mega:

Feel free to add any other relevant details such as __links, notes, screenshots__, here.

### CODE REVIEW

**Comment by upadhyeammit** [comment]
I have few questions,
1. We are using fallback value for unleash flag, this means that if unleash is enabled and for some reason unleash service does not response then we fallback to some value. Now this value is True. This means if unleash is enabled anyhow we are going to consider its enabled even when the unleash service itself is not able to respond. This is what I can see inventory is following as well. Is this fine?
2. Inventory has bypassed the unleash in the test env, however for us at least I want to execute the groups related unit tests and so I have not bypassed it. This is why we are mocking the unleash enablement. Is this fine?
3. Locally by default unleash has been bypassed. Is this fine?
4. If unleash is available and for some reasons rbac ended up not giving us the group details, in that case the current scenario is we show the systems which are not in any group, is this acceptable?

**Comment by PreetiW** [comment]
@upadhyeammit do we know why the PR checks are failing or do we want QE to look into it?

**Comment by upadhyeammit** [comment]
> @upadhyeammit do we know why the PR checks are failing or do we want QE to look into it?

Its failing for cache dir permissions, I am having a look!

**Comment by upadhyeammit** [comment]
> > @upadhyeammit do we know why the PR checks are failing or do we want QE to look into it?
> 
> Its failing for cache dir permissions, I am having a look!

I removed it, its not mandatory however I was expecting that it would work

https://docs.getunleash.io/Flask-Unleash/

**Comment by upadhyeammit** [comment]
> > > @upadhyeammit do we know why the PR checks are failing or do we want QE to look into it?
> > 
> > 
> > Its failing for cache dir permissions, I am having a look!
> 
> I removed it, its not mandatory however I was expecting that it would work
> 
> https://docs.getunleash.io/Flask-Unleash/

It seems this doesn't work either, now its creating it under /.cache, checking with appsre

**Comment by PreetiW** [comment]
@upadhyeammit can you please rebase this branch?

File: ros/api/common/add_group_filter.py [hunk 1]
```diff
@@ -1,12 +1,24 @@
...
+
+logger = get_logger(__name__)
 
 
 def group_filtered_query(query):
-    if hasattr(request, "host_groups"):
-        host_groups = [gid for gid in request.host_groups if gid is not None]
-        if len(host_groups) >= 1:
-            query = query.filter(System.groups[0]['id'].astext.in_(host_groups) | (System.groups == '[]'))
+    if get_flag_value(FLAG_INVENTORY_GROUPS):
```
Comment (PreetiW) [inline ros/api/common/add_group_filter.py:10]: @upadhyeammit according to me the feature flag checks should be at the API level as well and for RBAC if feature flag is enabled it should follow the condition based on RBAC response else we should skip adding any filter to the query is what I feel from the functionality point of view so that it will return all the systems and work like before. Let me know your thoughts!

Comment (PreetiW) [reply ros/api/common/add_group_filter.py:10]: What I mean by API level is that if feature flag is disabled that group info should not come from the Systems API response and filtering and sorting on `group_name` should not work.

File: pyproject.toml [hunk 2]
```diff
@@ -27,6 +27,7 @@ watchtower = ">=1.0.0"
 boto3 = ">=1.16.13"
 botocore = ">=1.19.13"
 flask-caching = "*"
+flask-unleash = "^2.0.0"
```
Comment (PreetiW) [inline pyproject.toml:30]: @upadhyeammit is inventory also using the same package? I am asking as it looks like t[he package is not maintained recently.](https://github.com/Unleash/Flask-Unleash)

File: clowdapp.yaml [hunk 3]
```diff
@@ -305,3 +350,13 @@ parameters:
...
+
+# Feature flags
+- description: Unleash secret name
+  name: UNLEASH_SECRET_NAME
+  value: bypass
+- description: Unleash API url
+  name: UNLEASH_URL
+- description: disable Unleash (feature flags), defaulting to fallback values
+  name: BYPASS_UNLEASH
+  value: 'false'
```
Comment (PreetiW) [inline clowdapp.yaml:362]: @upadhyeammit can you please explain a bit about the changes we did here for my understanding?

**Review by PreetiW — CHANGES_REQUESTED** [review]
Thank you for creating PR @upadhyeammit, I have left a few comments. Let me know if any questions, overall PR looks good 💯

**Comment by PreetiW** [comment]
@upadhyeammit here are my thoughts 💯 

> I have few questions,
> 
> 1. We are using fallback value for unleash flag, this means that if unleash is enabled and for some reason unleash service does not response then we fallback to some value. Now this value is True. This means if unleash is enabled anyhow we are going to consider its enabled even when the unleash service itself is not able to respond. This is what I can see inventory is following as well. Is this fine?  **According to me Yes 👍🏻** 
> 2. Inventory has bypassed the unleash in the test env, however for us at least I want to execute the groups related unit tests and so I have not bypassed it. This is why we are mocking the unleash enablement. Is this fine? **+1 with this**
> 3. Locally by default unleash has been bypassed. Is this fine? **+1 with this, I don't see any need of using unleash locally**
> 4. If unleash is available and for some reasons rbac ended up not giving us the group details, in that case the current scenario is we show the systems which are not in any group, is this acceptable? **I have added comments regarding how I think we need to use unleash feature flag, have a look and lets continue the discussion in that thread.**

File: ros/api/common/add_group_filter.py (hunk 1 above)
Comment (upadhyeammit) [reply ros/api/common/add_group_filter.py:10]: As per new changes if unleash is not enabled then we return empty groups array instead of completely getting rid of groups field from the response. This is to keep in sync with openapi spec

File: ros/api/v1/hosts.py [hunk 4]
```diff
@@ -257,7 +260,7 @@ def build_sort_expression(self, order_how, order_method):
             return (sort_order(PerformanceProfile.report_date),
                     asc(PerformanceProfile.system_id))
 
-        if order_method == 'group_name':
+        if order_method == 'group_name' and get_flag_value(FLAG_INVENTORY_GROUPS):
```
Comment (PreetiW) [inline ros/api/v1/hosts.py:263]: @upadhyeammit shall we revert this change similar to above group_name filter?

File: ros/api/v1/hosts.py [hunk 5]
```diff
@@ -319,6 +322,8 @@ def get(self, host_id):
             record['os'] = system.deserialize_host_os_data
             record['number_of_recommendations'] = profile.number_of_recommendations
             record['psi_enabled'] = profile.psi_enabled
+            if not get_flag_value(FLAG_INVENTORY_GROUPS):
+                record['groups'] = []
```
Comment (PreetiW) [inline ros/api/v1/hosts.py:326]: @upadhyeammit can you please share why we added this code? If I am not wrong this is for performance profile? If yes I am not sure if we need group info there.

File: tests/test_api_endpoints.py [hunk 6]
```diff
@@ -115,32 +117,40 @@ def test_system_groups(auth_token, db_setup, db_create_account, db_create_system
...
+        "id": "12345678-fe1b-4191-8408-cbadbd47f7a3",
         "name": "ros-test-3"
     }]
     db.session.commit()
 
     with app.test_client() as client:
+        mock_enable_rbac(mocker)
+        mock_rbac(get_rbac_mock_file("mock_rbac_returns_groups_including_example_group.json"), mocker)
+
+        mock_unleash_hbi_flag_enabled(mocker)
```
Comment (PreetiW) [inline tests/test_api_endpoints.py:129]: @upadhyeammit is it possible to add a test case when unleash flag is false and check for default values for groups in that case?

**Review by PreetiW — COMMENTED** [review]
@upadhyeammit Thank you for making the changes, just added minor comments on the PR. Rest is looking good 💯

File: ros/api/v1/hosts.py (hunk 4 above)
Comment (upadhyeammit) [reply ros/api/v1/hosts.py:263]: yeah!

Comment (PreetiW) [reply ros/api/v1/hosts.py:263]: @upadhyeammit I agree we are not assigning anything to the order method, I just felt that checking should be done after feature flag checking. But if you are okay with this, I am okay as well 👍🏻

File: ros/api/v1/hosts.py (hunk 5 above)
Comment (upadhyeammit) [reply ros/api/v1/hosts.py:326]: I think then maybe we don't even need the Groups column for Executive report?

Comment (PreetiW) [reply ros/api/v1/hosts.py:326]: @upadhyeammit I don't see any use case ATM where it will be useful in Executive report so I feel we dont need it there. But let me know if you can see any usecase.

Comment (upadhyeammit) [reply ros/api/v1/hosts.py]: I removed it from the pprofile.

**Review by PreetiW — APPROVED** [review]
Thank you for changes @upadhyeammit , LGTM 🚀
